src/wrapper/jnienv.rs

```diff
@@ -1932,6 +1966,192 @@ impl<'a> JNIEnv<'a> {
+    /// period of time before it calls release_primitive_array_critical. We must treat the code
+    /// inside this pair of functions as running in a "critical region." Inside a critical region,
+    /// native code must not call other JNI functions, or any system call that may cause the
+    /// current thread to block and wait for another Java thread. (For example, the current
+    /// thread must not call read on a stream being written by another Java thread.)
+    ///
+    /// These restrictions make it more likely that the native code will obtain an uncopied version
+    /// of the array, even if the VM does not support pinning. For example, a VM may temporarily
+    /// disable garbage collection when the native code is holding a pointer to an array obtained
+    /// via get_primitive_array_critical.
+    ///
+    /// Note that get_primitive_array_critical might still make a copy of the array if the VM
+    /// internally represents arrays in a different format. Therefore, we need to check its
+    /// return value against NULL for possible out of memory situations.
+    pub fn get_primitive_array_critical(
+        &self,
+        array: jarray,
+        is_copy: &mut jboolean,
+    ) -> Result<*mut c_void> {
+        non_null!(array, "get_primitive_array_critical array argument");
+        let res = jni_non_void_call!(self.internal, GetPrimitiveArrayCritical, array, is_copy);
+        Ok(res)
+    }
+
+    /// Release elements of the given array.
+    ///
+    /// See get_primitive_array_critical for a discussion on how these functions can be used.
+    pub fn release_primitive_array_critical(
+        &self,
+        array: jarray,
+        elems: &mut c_void,
+        mode: ReleaseMode,
+    ) -> Result<()> {
+        non_null!(array, "release_primitive_array_critical array argument");
+        jni_void_call!(
+            self.internal,
+            ReleasePrimitiveArrayCritical,
+            array,
+            elems,
+            mode as i32
+        );
+        Ok(())
+    }
+
+    /// Commit elements of the given primitive array.
+    ///
+    /// This function has no effect if elems is not a copy of the elements in array. Otherwise,
+    /// this function copies back the content of the array (and does not free the elems buffer).
+    pub fn commit_primitive_array_critical(
+        &self,
+        array: jbyteArray,
+        elems: &mut c_void,
+    ) -> Result<()> {
+        non_null!(array, "commit_primitive_array_critical array argument");
+        jni_void_call!(
+            self.internal,
+            ReleasePrimitiveArrayCritical,
+            array,
+            elems,
+            jni_sys::JNI_COMMIT
+        );
+        Ok(())
+    }
+
+    /// Return an AutoPrimitiveArray of the given Java primitive array.
+    ///
+    /// The result is valid until the corresponding AutoPrimitiveArray object goes out of scope,
+    /// when the release happens automatically according to the mode parameter.
```

Comment:
Do I understand correctly that this API might make the unsuspecting user non-intentionally break the conditions of "GetCritical" by using this method in a big scope: get for as short period as possible, release as soon as possible? E.g., 

```
var array = GetPrimitiveArrayCriticalAuto();
// Use array

// Do lots of other, potentially, long-running, stuff

// Finally, release it when it goes out of scope
```

So the questions are: 
- Shall we provide the auto flavour for critical arrays?
- If we do, shall we advise to consider explicit std::mem::drop if they are acquired in biggish scopes?

Reply:
Good points. I think we should provide the API, and specifically point to the dangers / limitations of Critical functions. `std::mem::drop` can be used, as you mention, but also scopes. I'll mention both.

Reply:
Perfect, thanks!